internal/ingress/controller/template/template.go

```diff
@@ -1672,3 +1673,45 @@ func convertGoSliceIntoLuaTable(goSliceInterface interface{}, emptyStringAsNil b
 		return "", fmt.Errorf("could not process type: %s", kind)
 	}
 }
+
+// buildOriginRegex returns an origin as a regex
+func buildOriginRegex(origin string) string {
+	if !strings.HasPrefix(origin, "*") {
+		return origin
+	}
+
+	origin = strings.Replace(origin, "*.", "", 1)
+	parts := strings.Split(origin, ".")
+
+	return "(" + strings.Join(parts, ".") + ")"
+}
+
+// buildCorsOriginRegex builds the regex string required by nginx
+func buildCorsOriginRegex(origins string) string {
+	var originSplitList = strings.Split(origins, ",")
+	if searchCorsOrigins(originSplitList, "*") {
+		return "set $http_origin *;\nset $cors 'true';"
+	}
+
+	var originsRegex string = "if ($http_origin ~* ("
+	for i, origin := range originSplitList {
+		originTrimmed := strings.TrimSpace(origin)
+		if len(originTrimmed) > 0 {
+			originsRegex = originsRegex + regexp.QuoteMeta(buildOriginRegex(originTrimmed))
+			if i != len(originSplitList)-1 {
+				originsRegex = originsRegex + "|"
+			}
+		}
+	}
+	originsRegex = originsRegex + ")$ ) { set $cors 'true'; }"
```

Comment:
I was reading through the `Origin` [header docs](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Origin#syntax) and it has a provision for an optional port. The `$` sign here may cause issues when that happens. 

I think that is worth capturing in the regex and adding a test perhaps

Reply:
You have a point. The problem is normally, origin is set by the requester and nginx will create the variable based on the header see docs here: $http_name

Therefore, the origin itself will contain the port and should be able to evaluate it. origin.com:9999

I can probably add a this if necessary with a test to ensure this is the case, if there isn't any. We'll see what other people would like.

Reply:
Yes this is what I meant

Reply:
I see what you meant after clearing up on slack :)
I'll definitely add a test.

Tested with:

```bash
christopher@chris-linux-lp:~/personal_code/github/ingress-nginx$ kc -n echoserver describe ingress 
Name:             echoserver
Namespace:        echoserver
Address:          10.96.23.13
Default backend:  default-http-backend:80 (<error: endpoints "default-http-backend" not found>)
Rules:
  Host              Path  Backends
  ----              ----  --------
  echo.example.com  
                    /   echoserver:80 (10.244.0.10:8080,10.244.0.8:8080,10.244.0.9:8080)
Annotations:        nginx.ingress.kubernetes.io/cors-allow-origin: http://test.com,http://test.com:9999,http://google.com
                    nginx.ingress.kubernetes.io/enable-cors: true
Events:
  Type    Reason  Age                    From                      Message
  ----    ------  ----                   ----                      -------
  Normal  Sync    2m28s (x2 over 2m39s)  nginx-ingress-controller  Scheduled for sync
```

```curl
curl -H "Origin: http://google.com:9000" -H "Host: echo.example.com" localhost:80 -v
```

returns a 200 OK response however, there are no CORS headers.
I'll add an optional port section